askup/views.py

```diff
@@ -1 +1,202 @@
+        user = self.request.user
+
+        if user.is_superuser:
+            return Qset.objects.filter(parent_qset_id=pk).order_by('name')
+        if user.is_authenticated():
+            return Qset.objects.filter(parent_qset=pk, top_qset__users=user.id).order_by('name')
+        else:
+            return []
+
+
+class QsetView(generic.ListView):
+    """Handles the Qset list view (subsets only/questions only/mixed)."""
+
+    model = Qset
+
+    def get_template_names(self):
+        """
+        Get template names to use in the view.
+
+        Overriding the get_template_names of generic.ListView
+        """
+        if self._current_qset_type == 1:
+            return ['askup/qset_subsets_only.html']
+        elif self._current_qset_type == 2:
+            return ['askup/qset_questions_only.html']
+        else:
+            return ['askup/qset_mixed.html']
+
+    def get_context_data(self, **kwargs):
+        """
+        Get context data for the list view.
+
+        Overriding the get_context_data of generic.ListView
+        """
+        current_qset = get_object_or_404(Qset, pk=self.kwargs.get('pk'))
+        self._current_qset_type = current_qset.type
+        context = super().get_context_data(**kwargs)
+
+        if self._current_qset_type == 2:
+            # Clear the qsets queryset if rendering the "questions only" Qset
+            context['object_list'] = []
+
+        if self._current_qset_type == 1:
+            # Clear the questions queryset if rendering the "qsets only" Qset
+            context['questions_list'] = []
+        else:
+            context['questions_list'] = Question.objects.filter(qset_id=self.kwargs.get('pk'))
+
+        context['main_title'] = current_qset.name
+        context['current_qset'] = current_qset
+        context['current_qset_name'] = current_qset.name
+        context['current_qset_id'] = current_qset.id
+        context['is_admin'] = self.request.user.is_superuser
+        context['is_teacher'] = 'Teachers' in self.request.user.groups.values_list('name', flat=True)
+        context['is_student'] = 'Students' in self.request.user.groups.values_list('name', flat=True)
+        context['is_qset_creator'] = context['is_admin'] or context['is_teacher']
+        context['is_qset_allowed'] = current_qset.type in (0, 1)
+        context['is_question_creator'] = self.request.user.is_authenticated()
+        return context
+
+    def get_queryset(self):
+        """
+        Get queryset for the list view.
+
+        Overriding the get_queryset of generic.ListView
+        """
+        queryset = Qset.objects.filter(parent_qset_id=self.kwargs.get('pk'))
+        parent_qset = get_object_or_404(Qset, pk=self.kwargs.get('pk'))
+
+        if parent_qset:
+            self._parent_qset = parent_qset
+
+        return queryset
+
+
+class QuestionView(generic.DetailView):
+    """Handles the Question detailed view."""
+
+    model = Question
+    template_name = 'askup/question.html'
```

Comment:
I would rename `question.html` to `question_detail.html` and remove L150

Reply:
done